ir/expression.cpp

```diff
@@ -137,3 +137,45 @@ IR::Constant IR::Constant::operator-() const { return IR::Constant(-value); }
 IR::Constant IR::Constant::GetMask(unsigned width) {
     return (IR::Constant(1) << width) - IR::Constant(1);
 }
+
+const IR::Constant *IR::Constant::get(const IR::Type *t, big_int v, Util::SourceInfo si) {
+    // Only cache bits with width lower than 16 bit to restrict the size of the cache.
+    const auto *tb = t->to<Type_Bits>();
+    if (t->width_bits() > 16 || tb == nullptr) {
+        return new IR::Constant(si, t, v);
+    }
+    // Constants are interned. Keys in the intern map are pairs of types and values.
+    using key_t = std::tuple<int, RTTI::TypeId, bool, big_int>;
+    static std::map<key_t, const Constant *> CONSTANTS;
+
+    auto *&result = CONSTANTS[{tb->width_bits(), t->typeId(), tb->isSigned, v}];
+    if (result == nullptr) {
+        result = new Constant(si, tb, v);
+    }
+
+    return result;
+}
+
+const IR::BoolLiteral *IR::BoolLiteral::get(bool value, const Util::SourceInfo &si) {
+    // Boolean literals are interned.
+    static std::map<bool, const IR::BoolLiteral *> LITERALS;
```

Comment:
this does not look like a job for a map :-). There are just 2 possible values.

Reply:
Once we are taking SourceInfo into account we may have to resort to a map again, but for now I am just declaring two static values.